EX2Linux.cpp

```diff
@@ -0,0 +1,174 @@
+/**
+ * this function recieves vector of char* and frees the memory allocated for it
+*/
+void freeArgs(std::vector<char*>& args) {
+    for (char* arg : args) {
+        delete[] arg;
+    }
+}
+
+//------------------------------------------------------------------------------------------
+/**
+ * this function recieves the first cell of the vector char* which is the command, a vector of 
+ * arguments which represent the arguments, a boolean value to see if the command must run on 
+ * bg or fg, and a vector of pid_t whihc will store all processes ran in bg. the function's 
+ * goal is to execute the command, the execvp uses PATHS to search for the command and executes it
+ * if the path is not found an error is displayed to user, it first creates a process:
+ * CHILD PROCESS: uses the execvp function to execute the command with the following arguments 
+ * PARENT PROCESS: checks first if the process must run in bg or fg, if fg, it waits for the 
+ *                  child process to finish, if bg, it does not need to wait the process keeps
+ *                  running on bg, and appends the process to the vector of bgprocesses.
+ * the function as well handles errors if any failures occur.
+*/
+void executeCommand(const std::string& command, std::vector<char*>& args, bool isBackground,
+                    std::vector<pid_t>& backgroundProcesses) {
+    pid_t pid = fork();
+    if(pid == 0){
+        //CHILD
+        if (execvp(command.c_str(), args.data()) == -1) {
+            perror("Command not found.");
+            exit(1);
+        }
+    }
+    else if(pid > 0){
+        //PARENT
+        if (isBackground) {
+            backgroundProcesses.push_back(pid); 
+        } else {
+            waitpid(pid, nullptr, 0);
+        }
+    }
+    else{
+        perror("Creating a process failed.");
+        exit(1);
+    }
+}
+
+//------------------------------------------------------------------------------------------
+/**
+ * this function receives a string which represents the input the user provided, and checks 
+ * if the user wants to run the command on bg, it does so by checking if the user provided 
+ * & at the end of the input, if so it returns true, meaning the command must run in bg, 
+ * and false otherwise.
+*/
+bool isBackgroundCommand(const std::string& input) {
+    return input.back() == '&';
+}
+
+//------------------------------------------------------------------------------------------
+/**
+ * this function recieves a vector of process IDs, consisting of every vector running on 
+ * background. the function displays the id of the process running on bg, as well as 
+ * the status. it uses signals to retrieve this type of data.
+*/
+void myjobs(const std::vector<pid_t>& backgroundProcesses) {
+    std::cout << "Background processes:" << std::endl;
+    std::cout << "PID\tExit Status\tStatus" << std::endl;
+
+    // Check the status of each background process
+    for (pid_t pid : backgroundProcesses) {
```

Comment:
What about cleaning up processes that terminated? There is not point in showing the entire background process history forever. IMO after the process terminated it should show up once and then be deleted from the list. Otherwise the list might grow forever..